st2actions/tests/integration/test_resultstracker.py

```diff
@@ -58,16 +110,168 @@ def test_querier_started(self):
+
+        with mock.patch.object(
+                querier.__class__, '_update_action_results',
+                mock.MagicMock(side_effect=Exception('Mock update exception.'))):
+            tracker._bootstrap()
+            eventlet.sleep(1)
+
+            exec_id = str(ResultsTrackerTests.states['state1.yaml'].execution_id)
+            exec_db = LiveAction.get_by_id(exec_id)
+            self.assertDictEqual(exec_db.result, {})
+
+            exec_id = str(ResultsTrackerTests.states['state2.yaml'].execution_id)
+            exec_db = LiveAction.get_by_id(exec_id)
+            self.assertDictEqual(exec_db.result, {})
+
+            tracker.shutdown()
+
+        # Ensure deletes are not called.
+        querier._delete_state_object.assert_not_called()
+
+    @mock.patch.object(
+        Action, 'get_by_ref', mock.MagicMock(return_value='foobar'))
+    def test_execution_cancellation(self):
+        tracker = results_tracker.get_tracker()
+        querier = tracker.get_querier('test_querymodule')
+        querier._delete_state_object = mock.Mock(return_value=None)
+        querier._invoke_post_run = mock.Mock(return_value=None)
+
+        with mock.patch.object(
+                querier.__class__, 'query',
+                mock.MagicMock(return_value=(action_constants.LIVEACTION_STATUS_CANCELED, {}))):
+            tracker._bootstrap()
+            eventlet.sleep(1)
+
+            exec_id = str(ResultsTrackerTests.states['state1.yaml'].execution_id)
+            exec_db = LiveAction.get_by_id(exec_id)
+            self.assertDictEqual(exec_db.result, {})
+
+            exec_id = str(ResultsTrackerTests.states['state2.yaml'].execution_id)
+            exec_db = LiveAction.get_by_id(exec_id)
+            self.assertDictEqual(exec_db.result, {})
+
+            tracker.shutdown()
+
+        # Ensure deletes are called.
+        self.assertEqual(2, querier._delete_state_object.call_count)
+
+        # Ensure invoke_post_run is not called.
+        querier._invoke_post_run.assert_not_called()
+
+    @mock.patch.object(
+        Action, 'get_by_ref', mock.MagicMock(return_value=None))
+    def test_action_deleted(self):
+        tracker = results_tracker.get_tracker()
+        querier = tracker.get_querier('test_querymodule')
+        querier._invoke_post_run = mock.Mock(return_value=None)
+
+        tracker._bootstrap()
+        eventlet.sleep(1)
+
+        exec_id = str(ResultsTrackerTests.states['state1.yaml'].execution_id)
+        exec_db = LiveAction.get_by_id(exec_id)
+        self.assertTrue(exec_db.result['called_with'][exec_id] is not None,
+                        exec_db.result)
+
+        exec_id = str(ResultsTrackerTests.states['state2.yaml'].execution_id)
+        exec_db = LiveAction.get_by_id(exec_id)
+        self.assertTrue(exec_db.result['called_with'][exec_id] is not None,
+                        exec_db.result)
+
+        tracker.shutdown()
+
+        # Ensure state objects are deleted
+        self.assertRaises(
```

Comment:
Should we also verify somewhere at the top that this object actually exists in the database in the first place?